src/shared/sam/sync.ts

```diff
@@ -543,9 +572,15 @@ export function registerSync() {
     ): Promise<SamSyncResult> {
         telemetry.record({ syncedResources: deployType === 'infra' ? 'AllResources' : 'CodeOnly' })
 
-        const connection = Auth.instance.activeConnection
-        if (connection?.type !== 'iam') {
-            throw new ToolkitError('Syncing SAM applications requires IAM credentials', { code: 'NoIAMCredentials' })
+        let connection = Auth.instance.activeConnection
+        if (connection?.type !== 'iam' || connection?.state !== 'valid') {
+            await promptReauth(connection)
+            connection = Auth.instance.activeConnection
+            if (connection?.type !== 'iam' || connection?.state !== 'valid') {
+                throw new ToolkitError('Syncing SAM applications requires IAM credentials', {
+                    code: 'NoIAMCredentials',
+                })
+            }
         }
```

Comment:
- If `promptReauth` returned the connection (or returned early if it's already valid), this logic could be greatly simplified.
- Also allows `const`. Removing `const` is usually a signal that logic could be improved.
- Renaming it to `getAuthOrPrompt` could make it clearer.


```suggestion
        const connection = await promptReauth(connection)
        if (!connection) {
            throw new ToolkitError('Syncing SAM applications requires IAM credentials', {
                code: 'NoIAMCredentials',
            })
        }
```